### PR TITLE
[WIP] New build system

### PR DESCRIPTION
**Pull request by agramfort**
this PR is a full review of the building system that:

- uses modern cmake syntax
- allows us automatic shipping of working binaries on all 3 systems (tested in linux, mac and windows) using travis and appveyor
- allows us to have continuous time of less than 10 mins per build (instead of 2h so 10x speed on CI)
- will allow us to make easily conda packages (still to do in a next PR)
- will allow us to make easily brew packages with openmp
- it significantly reduces the number of lines of cmake code and its maintenance
- no more sudo requirements inside cmake files

Things to do in subsequents PRs:

- activate clang analyzer and fix the issues it reports. Not done here to avoid touch C++ code.
- put back Python wrapping.
- put back CGAL

cc @papadop @massich @mclerc @eolivi 
 @m9h

### CODE REVIEW

File: cmake/utils/CheckCXXFeatures.cmake [hunk 1]
```diff
@@ -85,31 +85,31 @@ macro(CHECK_CXX_STANDARD_LIBRARY)
...
+    # CHECK_CXX_FEATURE(HAVE_MATH_FN_IN_NAMESPACE_STD      math_fn_in_namespace_std.cpp      "has C math functions in namespace std")
+    # CHECK_CXX_FEATURE(HAVE_MATH_ABSINT_IN_NAMESPACE_STD  math_absint_in_namespace_std.cpp  "has C math abs(integer type) in namespace std")
+    # CHECK_CXX_FEATURE(HAVE_COMPLEX_MATH_IN_NAMESPACE_STD complex_math_in_namespace_std.cpp "supports complex math functions are in namespace std")
+    # CHECK_CXX_FEATURE(HAVE_ISNAN_IN_NAMESPACE_STD        isnan_in_namespace_std.cpp        "has isnan function in namespace std")
+    CHECK_CXX_FEATURE(HAVE_ISNORMAL_IN_NAMESPACE_STD     isnormal_in_namespace_std.cpp     "has isnormal function in namespace std")
+    # CHECK_CXX_FEATURE(HAVE_STD                           have_std.cpp                      "supports ISO C++ standard library")
+    # CHECK_CXX_FEATURE(HAVE_STL                           have_stl.cpp                      "supports Standard Template Library")
+    # CHECK_CXX_FEATURE(HAVE_RUSAGE                        have_rusage.cpp                   "has getrusage() function")
```
Comment (agramfort) [inline cmake/utils/CheckCXXFeatures.cmake]: @papadop we commented out the test that we are not actually using.

File: apps/tools/CMakeLists.txt [hunk 2]
```diff
@@ -23,13 +23,13 @@ add_executable(om_register_squids register_squids.cpp)
...
-target_link_libraries(om_squids2vtk OpenMEEGMaths)
+target_link_libraries(om_squids2vtk OpenMEEGMaths OpenMEEG::OpenMEEG)
 
 add_executable(om_matrix_info matrix_info.cpp)
-target_link_libraries(om_matrix_info OpenMEEGMaths)
+target_link_libraries(om_matrix_info OpenMEEGMaths OpenMEEG::OpenMEEG)
 
 add_executable(om_matrix_convert matrix_convert.cpp)
-target_link_libraries(om_matrix_convert OpenMEEGMaths)
+target_link_libraries(om_matrix_convert OpenMEEGMaths OpenMEEG::OpenMEEG)
```
Comment (agramfort) [inline apps/tools/CMakeLists.txt]: @papadop here is an example of modern cmake.

File: OpenMEEGMaths/include/BlasLapackImplementations/OpenMEEGMathsvecLibConfig.h [hunk 3]
```diff
@@ -3,10 +3,8 @@
 
 #include <BlasLapackImplementations/FortranCInterface.h>
 
-extern "C" {
-    #include <cblas.h>
-    #include <clapack.h>
-}
+# include <Accelerate/Accelerate.h>
```
Comment (agramfort) [inline OpenMEEGMaths/include/BlasLapackImplementations/OpenMEEGMathsvecLibConfig.h]: this is the clean way on mac.

File: OpenMEEGMaths/include/BlasLapackImplementations/OpenMEEGMathsAtlasConfig.h [hunk 4]
```diff
@@ -3,8 +3,7 @@
 #include <BlasLapackImplementations/FortranCInterface.h>
 
 extern "C" {
-    #include <cblas.h>
-    #include <clapack.h>
+    #include <atlas/clapack.h>
```
Comment (agramfort) [inline OpenMEEGMaths/include/BlasLapackImplementations/OpenMEEGMathsAtlasConfig.h]: this seems to be clean on travis. Maybe to be tested on your fedora.

**Comment by agramfort** [comment]
CIs are green ! 🥇 🥇 🥇 🥇 🥇 🥇 !

**Comment by sagitter** [comment]
Does this new build system work with Openblas? (BLA_VENDOR=ATLAS ?)

```
+ /usr/bin/unzip -qq /home/sagitter/rpmbuild/SOURCES/openmeeg-6c7194882d3e1aee60489790b3cf9800d74186e5.zip
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd openmeeg-6c7194882d3e1aee60489790b3cf9800d74186e5
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ exit 0
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.cJ2nhb
+ umask 022
+ cd /home/sagitter/rpmbuild/BUILD
+ cd openmeeg-6c7194882d3e1aee60489790b3cf9800d74186e5
+ mkdir -p build
+ pushd build
~/rpmbuild/BUILD/openmeeg-6c7194882d3e1aee60489790b3cf9800d74186e5/build ~/rpmbuild/BUILD/openmeeg-6c7194882d3e1aee60489790b3cf9800d74186e5
+ export 'LDFLAGS=-Wl,-z,relro -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -Wl,--as-needed'
+ LDFLAGS='-Wl,-z,relro -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -Wl,--as-needed'
+ export PYTHON_DISALLOW_AMBIGUOUS_VERSION=0
+ PYTHON_DISALLOW_AMBIGUOUS_VERSION=0
+ CFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables'
+ export CFLAGS
+ CXXFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables'
+ export CXXFLAGS
+ FFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -I/usr/lib64/gfortran/modules'
+ export FFLAGS
+ FCFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -I/usr/lib64/gfortran/modules'
+ export FCFLAGS
+ LDFLAGS='-Wl,-z,relro -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -Wl,--as-needed'
+ export LDFLAGS
+ /usr/bin/cmake -DCMAKE_C_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_Fortran_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_INSTALL_PREFIX:PATH=/usr -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib64 -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DLIB_SUFFIX=64 -DBUILD_SHARED_LIBS:BOOL=ON .. -DCMAKE_BUILD_TYPE=Release -DUSE_PROGRESSBAR=ON -DBUILD_DOCUMENTATION:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=ON -DBUILD_SHARED_LIBS_OpenMEEG:BOOL=ON -DBUILD_SHARED_LIBS_matio:BOOL=ON -DBUILD_SHARED_LIBS_zlib:BOOL=ON -DCMAKE_SKIP_INSTALL_RPATH:BOOL=YES -DBLA_VENDOR=ATLAS -DLAPACK_openblas_LIBRARY:FILEPATH=/usr/lib64/libopenblas.so -DBLAS_openblas_LIBRARY:FILEPATH=/usr/lib64/libopenblas.so -DENABLE_PYTHON:BOOL=ON -DBUILD_TESTING:BOOL=ON -DBUILD_TOOLS:BOOL=ON -DENABLE_PACKAGING:BOOL=OFF -DSKIP_GITHUB_TESTS:BOOL=OFF -DUSE_CGAL:BOOL=OFF -DUSE_GIFTI:BOOL=ON -DUSE_OMP:BOOL=ON -DUSE_SYSTEM_hdf5:BOOL=ON -DUSE_SYSTEM_matio:BOOL=ON -DBLASLAPACK_IMPLEMENTATION:STRING=OpenBLAS -DUSE_SYSTEM_zlib:BOOL=ON -DLapacke_LIB:FILEPATH= -DUSE_VTK:BOOL=OFF -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON
-- The C compiler identification is GNU 7.3.1
-- The CXX compiler identification is GNU 7.3.1
-- Check for working C compiler: /usr/lib64/ccache/cc
-- Check for working C compiler: /usr/lib64/ccache/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/lib64/ccache/c++
-- Check for working CXX compiler: /usr/lib64/ccache/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check whether the compiler implements namespaces
-- Check whether the compiler implements namespaces - found
-- Check whether the compiler has isnormal function in namespace std
-- Check whether the compiler has isnormal function in namespace std - found
CMake Error at /usr/share/cmake/Modules/FindBLAS.cmake:699 (message):
  A required library with BLAS API not found.  Please specify library
  location.
Call Stack (most recent call first):
  cmake/thirdParties.cmake:32 (find_package)
  CMakeLists.txt:20 (include)
```

**Comment by agramfort** [comment]
It does at least on ubuntu and Debian cf Travis file

**Comment by sagitter** [comment]
I don't see OpenBLAS libraries in [log file](https://api.travis-ci.org/v3/job/353935365/log.txt)

**Comment by agramfort** [comment]
True openblas was not ported yet. One reason is that cmake find blas does not work out of the box with openblas 

We will add it soon after validation here https://github.com/massich/findopenblas_cmake

**Pull request closed without merging**